libs/x-content/src/main/java/org/elasticsearch/xcontent/DotExpandingXContentParser.java

```diff
@@ -19,11 +19,12 @@
  */
 public class DotExpandingXContentParser extends FilterXContentParser {
 
-    private static class WrappingParser extends DelegatingXContentParser {
+    private static class WrappingParser extends FilterXContentParser {
 
         final Deque<XContentParser> parsers = new ArrayDeque<>();
 
         WrappingParser(XContentParser in) throws IOException {
+            super(null);
```

Comment:
This is the bit that worries me, because it's too easy to write `in.delegetedMethod()` instead of `delegate().delegatedMethod()` and get weird NPEs.

Reply:
Maybe we could keep both, but FilterXContentParser could extend DelegatingXContentParser and have a final `delegate()` method that returns `in`?

Reply:
eheh yea I agree this is not great, the null value in particular. i also tried a Supplier but it's messy because it can't be provided in the constructor. I kept both in the beginning but I was not happy with their javadocs and their confusing name. `in` can be made private though so it can't be retrieved directly?

Reply:
I have pushed an update: I find it overkill to have a new base class for one single consumer. The downside is that we need to accept a null parser, and we have no good way to assert that when we do so, the corresponding subclass implements its own delegate(). Though this type of stuff should be a fundamental error that is easily caught in early testing?

Reply:
With another recent change I ended up creating an additional subclass needing a dynamic delegate, so that became a good reason to follow your advice :) 

I tweaked naming slightly: FilterXContentParser is the one that allows to override delegate, while FilterXContentParserWrapper takes the delegate as constructor argument.